srml/slashing/src/lib.rs

```diff
@@ -0,0 +1,211 @@
+//!			// let balance = T::Currency::free_balance(who);
+//!			// let slash = balance / severity;
+//!			// T::Currency::on_slash(who, balance);
+//!		}
+//! }
+//!
+//!	struct SlashingWrapper<T>(PhantomData<T>);
+//!
+//!	impl<T: Trait> Slashing<T::AccountId> for SlashingWrapper<T> {
+//!		type Slash = Balance<T>;
+//!
+//!		fn slash<M: Misconduct>(
+//!			misbehaved: &[T::AccountId],
+//!			total_validators: u64,
+//!			misconduct: &mut M
+//!		) -> u8 {
+//!			misconduct.on_misconduct(misbehaved.len() as u64, total_validators);
+//!			let severity = misconduct.severity();
+//!
+//!			for who in misbehaved {
+//!				Self::Slash::on_slash::<M>(who, severity);
+//!			}
+//!
+//!			severity.as_misconduct_level()
+//!		}
+//!	}
+//!
+//! fn main() {
+//!		let unresponsive = Unresponsive;
+//!		let misbehaved_validators = vec![1, 2, 3, 4, 5, 6];
+//!		let total_validators = 100;
+//!		// MyModule is type that implements `Trait`
+//!		// SlashingWrapper::<MyModule>::slash_on_checkpoint(&misbehaved_validators, total_validators, &unresponsive);
+//! }
+//! ```
+
+use parity_codec::Codec;
+use primitives::traits::{SimpleArithmetic, MaybeSerializeDebug};
+
+/// Pre-defined types
+pub mod misconduct;
+
+mod types;
+
+pub use types::Fraction;
+
+// The specification specifices four different misconduct levels:
+//		1) Slashing: 0 <= x <= 0.001
+//		2) Slashing: 0.001 < x <= 0.01
+//		3) Slashing: 0.01 < x <= 0.1
+//		4) Slashing: 0.1 < x <= 1.0
+type MisconductLevel = u8;
+
+/// Base trait for representing misconducts
+pub trait Misconduct {
+	/// Severity represented as a fraction
+	type Severity: SimpleArithmetic + Codec + Copy + MaybeSerializeDebug + Default + Into<u128>;
+
+	/// Valid `Window` size in `Sessions` for the given misconduct
+	// (niklasad1): an `Era` is hard-coded to 6 sessions?!
+	const WINDOW_SIZE: u32;
+
+	/// Estimate severity level on the misconduct and store the state
+	///
+	// TODO(niklasad1): shall `num_misbehaved` & `num_validators` be generic?!
+	fn on_misconduct(&mut self, num_misbehaved: u64, num_validators: u64);
+
+	/// Get severity level
+	///
+	/// Returns `Some(severity)` if at least k samples are received otherwise `None`
+	fn severity(&self) -> Fraction<Self::Severity>;
+
+}
+
+/// Slashing interface
+pub trait OnSlashing<AccountId> {
+	/// Slash validator `who` based on severity_level `severity`
+	fn on_slash<M: Misconduct>(who: &AccountId, severity: Fraction<M::Severity>);
+}
+
+/// Slashing wrapper interface on top of `OnSlashing`
+pub trait Slashing<AccountId> {
+	/// Specify which `OnSlashing` implementation to use
+	type Slash: OnSlashing<AccountId>;
+
+	/// Attempt to slash a list of `misbehaved` validators
+	///
+	/// Returns the misconduct level for all misbehaved validators
+	// TODO(niklasad1): shall `total_validators` be generic?!
+	fn slash<M: Misconduct>(
+		misbehaved: &[AccountId],
+		total_validators: u64,
+		misconduct: &mut M
```

Comment:
It is not obvious to me how to keep track of the which the current session is and when it changes.

If we could make the assumption that `Slashing::slash` would be called at least once every session it would be quite straight-forward to implement and pass it to `Slash`